pkg/git/provider_bitbucket-server.go

```diff
@@ -46,22 +54,82 @@ func NewBitbucketServerProvider(baseURL string) (Provider, error) {
 	}, nil
 }
 
-func (bbs *bitbucketServer) Type() ProviderType {
-	return bbs.providerType
-}
-
 func (bbs *bitbucketServer) BaseURL() string {
 	urlClone := *bbs.apiURL
 	urlClone.Path = ""
 	urlClone.RawQuery = ""
 	return urlClone.String()
 }
 
-func (bbs *bitbucketServer) VerifyToken(ctx context.Context, tokenType TokenType, token string) error {
-	log.G(ctx).Debug("Skip verifying token for bitbucket, to be implemented later")
+func (bbs *bitbucketServer) SupportsMarketplace() bool {
+	return false
+}
+
+func (bbs *bitbucketServer) Type() ProviderType {
+	return bbs.providerType
+}
+
+func (bbs *bitbucketServer) VerifyRuntimeToken(ctx context.Context, token string) error {
+	return bbs.checkProjectAdminPermission(ctx, token)
+}
+
+func (bbs *bitbucketServer) VerifyUserToken(ctx context.Context, token string) error {
+	return bbs.checkRepoReadPermission(ctx, token)
+}
+
+func (bbs *bitbucketServer) checkProjectAdminPermission(ctx context.Context, token string) error {
+	username, err := bbs.getCurrentUsername(ctx, token)
+	if err != nil {
+		return err
+	}
+
+	urlPath := fmt.Sprintf("users/%s/repos", username)
+	body := &createRepoBody{
+		Name: "!invalid",
+	}
+	res, err := bbs.request(ctx, token, http.MethodPost, urlPath, body)
+	if err != nil {
+		return fmt.Errorf("failed checking Project admin permission: %w", err)
+	}
+	defer res.Body.Close()
+
+	if res.StatusCode != http.StatusBadRequest {
+		return errors.New("git-token is invalid or missing required \"Project admin\" scope")
+	}
+
+	return nil
+}
+
+func (bbs *bitbucketServer) checkRepoReadPermission(ctx context.Context, token string) error {
+	username, err := bbs.getCurrentUsername(ctx, token)
+	if err != nil {
+		return fmt.Errorf("failed checking Repo read permission: %w", err)
+	}
+
+	if username == "" {
+		return errors.New("personal-git-token is invalid")
+	}
+
 	return nil
 }
 
-func (bbs *bitbucketServer) SupportsMarketplace() bool {
-	return false
+func (bbs *bitbucketServer) getCurrentUsername(ctx context.Context, token string) (string, error) {
+	res, err := bbs.request(ctx, token, http.MethodHead, "application-properties", nil)
```

Comment:
HEAD to `application-properties` - this endpoint does not require any permission (or auth header). but any request with a valid token has the `X-AUSERNAME` response header with the user name in it.